hips/hip-00NN.md

```diff
@@ -0,0 +1,101 @@
+created: "2021-10-05"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Currently, Helm supports installing dependencies from various registries, however that requires the charts to be published there. There are use cases which are cumbersome when registries are involved, for example:
+
+- private charts for organizations/individuals who do not have the capacity to maintain the infrastructure required to run a private registry
+- testing work-in-progress charts in their umbrella charts
+
+## Motivation
+
+There are existing ways to achieve installation of charts without using a registry, however they are not very user-friendly and require additional tooling.
+
+- A Helm chart repository is effectively an `index.yaml` with links to downloads. Maintaining such a repository does create a burden of scripts and automation (e.g. Github Actions). This is not always feasible for smaller projects. It also does not really offer an obvious and readily available way of testing pre-releases.
+- For the testing use cases, charts can be packaged using `helm package`, however this does introduce manual steps and requires extra work to replicate in CI/CD scenarios.
+- There are several plugins available to solve this problem ([`aslafy-z/helm-git`](https://github.com/aslafy-z/helm-git), [`diwakar-s-maurya/helm-git`](https://github.com/diwakar-s-maurya/helm-git), [`sagansystems/helm-github`](https://github.com/sagansystems/helm-github)), however using plugins requires additional setup, which may not always be feasible (esp. in more complex team structures and cluster setups with advanced tooling, e.g. ArgoCD). An additional drawback is that the `Chart.yaml` does not provide a way to specify the plugin requirements, which leaves it up to the consumer to figure this out.
+
+## Rationale
+
+Installing dependencies from git is an established pattern in other ecosystems even when they are registry-based, e.g. npm (Node.js), pipenv (Python), bundler (Gem) have this option - it would make sense to have the behavior replicated in Helm.
+
+At least the npm and the pip ecosystems have already established a syntax of `vcs+protocol` for defining the dependency source, so it should be familiar to some users.
+
+One alternative to consider would be to exclude defining the vcs/protocol, similar to what Go does, esp. given that Helm is built using Go. This does limit the flexibility somewhat - while Go does allow adding a VCS qualifier at the end of the URL (allowing future support for other VCSs), it does not allow specifying the protocol, which means that the users might have to override the default protocol in their VCS configuration.
+
+## Specification
+
+The `Chart.yaml` should support the following format for `dependencies`:
+
+```
+dependencies:
+- name: "<dependency name>"
+  repository: "<protocol>://[<user>[:<password>]@]<hostname>[:<port>][:][/]<path>"
```

Comment:
I wonder if we should allow (read: explicitly disallow) username / password in `Chart.yaml` dependency URL? Charts are generally intended for distribution, and credentials in things intended for distribution isn't great. Helm might want to proactively protect users here. 

If we decide to allow. I think a lint rule would be desired.

Generally, the [git-credential](https://git-scm.com/docs/git-credential) process (as invoked by `git clone`) I think would be the preferred method for git to gain access to credentials I think.

Reply:
I'll admit I just copy/pasted this from `npm` documentation.

I agree that it should be removed: https://github.com/helm/community/pull/321/commits/888817fc276a6ab6a0415b71efaf7270a7f6a0a4

That said, I think it would just work, because whatever works for `git clone` should work in `repository`. Do you think it should be explicitly validated and forbidden, or is leaving it undocumented enough?

The implementation code which documents the usage does recommend using a credentials helper: https://github.com/dominykas/helm/blob/e4b4b5e1811af0e18f28a7e4efe3f4b01ebebf59/cmd/helm/dependency.go#L96-L101 (very early wip).

Reply:
> I'll admit I just copy/pasted this from npm documentation.

😆 

---

I think we should decide to do one of:
1. explicitly disallow (via validation in Helm; I think the validation would be trivial to implemt: `u := url.Parse() + u.User().Password() == ""`)
2. not document, and have a lint rule to warn the user they are potentially including sensitive credentials

HIP IMHO should document/detail which decision we decide

Reply:
Added a note to explicitly forbid u/p in the implementation: https://github.com/helm/community/pull/321/commits/7b4771b0976dadb778ae1ba39dd7c01ada5f0549 - it is indeed easier to relax constraints down the road, based on user feedback.